Essentials/src/com/earth2me/essentials/Jails.java

```diff
@@ -117,12 +118,31 @@ public void removeJail(final String jail) throws Exception {
     }
 
     @Override
+    @Deprecated
     public void sendToJail(final IUser user, final String jail) throws Exception {
+        CompletableFuture<Exception> eFuture = new CompletableFuture<>();
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        sendToJail(user, jail, eFuture, future);
+        if (!future.get()) {
+            throw eFuture.get();
+        }
```

Comment:
I'm fairly certain that if a 3rd party plugin calls this it will deadlock the server. Should have been more clear when talking about 'deadlocking' in discord.

I'm fairly certain this is the case - imagine the following:

1) sendToJail is called
2) the server has a 'cache miss', resulting in a chunk load.
3) but wait! we are using paper, so its a cool async chunk load 
4) server goes off and loads the chunk in a separate thread
5) Server tries to continue to 126 - which is blocked on a lock. Main thread is now dependent on the async thread completing (which is what we want, because this is maintaining legacy behavior)
6) Server is done 'loading' the chunk via IO, but now must load the chunk into the internal chunk map, which includes loading entities
7) Entities have to be loaded in sync because they have to figure out bounding box stuff which calls into big block methods which look at worlds at a whole, not just the chunk
8) I would assume Paper then waits for the next sync chunk to load the entities and declare the chunk 'loaded'

At this point, we have deadlocked because the sync thread is unable to progress in order to process the final chunk load, yet in order to progress the chunk must be loaded.

Reply:
Should be mitigated as of now